src/scenario.js

```diff
@@ -240,108 +223,88 @@ export class Scenario extends Debug {
-    while (this.#stepPointer < this.steps.length) {
-      const currentStep = this.steps[this.#stepPointer];
-
-      switch (currentStep.type) {
-        case kStepType.expect: {
-          const bufferValue = this.#buffer.out.shift();
-
-          this._compare(currentStep, bufferValue);
-          this.#next();
-
-          yield currentStep;
-          break;
-        }
-        case kStepType.exitCode: {
-          const actualCode = this.#buffer.code;
-
-          this._compare(currentStep, actualCode);
-          this.#next();
-
-          yield currentStep;
-          break;
-        }
-        case kStepType.expectError: {
-          const bufferValue = this.#buffer.err.shift();
-
-          this._compare(currentStep, bufferValue);
-          this.#next();
-
-          yield currentStep;
-          break;
-        }
-        case kStepType.input: {
-          this._writeInProc(currentStep.value);
-
-          currentStep.ok = true;
-          this.debug('input', currentStep.value);
-          this.#next();
-
-          await new Promise((resolve) => this.#pipe(resolve));
-
-          yield currentStep;
-          break;
-        }
-        default: {
-          throw new Error(`step ${currentStep.type} doesn't exist`);
-        }
-      }
-    }
-  }
-
-  /**
-   * Increase the step pointer
-   */
   #next() {
     this.#stepPointer++;
   }
 
-  /**
-   * Spawn the command
-   * @returns {Promise<void>} - resolve when the command is spawned and first outputs are received
-   */
+  get #currentStep() {
+    return this.steps.at(this.#stepPointer);
+  }
+
+  #resetTimer() {
+    clearTimeout(this.#timer);
+  }
+
+  #startTimer(done) {
+    this.#timer = setTimeout(done, this.#globalTimeout);
+  }
+
   async #spawnCommand() {
-    return new Promise((resolve) => {
-      const proc = spawn(this.#command, {
-        shell: true,
+    return new Promise((resolve, reject) => {
+      const context = {
+        done: resolve,
+        reject,
+      };
+      const process = new Process();
+
+      this.#startTimer(resolve);
+
+      process.spawn(this.#command);
+
+      process.on('spawn', (pid) => {
+        this.debug('spawn, pid:', pid);
+        this._proc = process;
       });
 
-      proc.on('spawn', () => {
-        this.debug('spawn, pid:', proc.pid);
-        this._proc = proc;
-        this.#pipe(resolve);
+      process.on('data', (line) => {
+        this.debug('data received: ', line);
+        this.#handleData(line, { ...context, isError: false });
       });
 
-      proc.on('exit', (code) => {
-        this.#buffer.code = code;
+      process.on('error', async (line) => {
+        this.debug('error received: ', line);
+        this.#handleData(line, { ...context, isError: true });
+      });
+
+      process.on('exit', async (code) => {
+        this.debug('exited with code: ', code);
+        this.#handleData(code, { ...context, isError: code != 0 });
       });
     });
   }
```

Comment:
I'd like to share something on this part of the code. 

What If we extend the responsibility of the Process class?

With the current implementation, I found that Process behaves like a reusable class (from other contexts than `clix`). I imagined something more like that:

```js
      const context = {
        resolve,
        reject,
        handleData
      };
      const process = new Process(context);
      
      this.#startTimer(resolve);

      process.spawn(this.#command);
      // EOF
```

... and let process call `handeData` with the proper arguments, then on the `Process` side we don't have to care about subscribers stuff for example.
In a nutshell, I'd like to have a Process in the sense of `clix` and not a Process in the general-purpose sense. (we could rename it)

I could give it a try and then we could take a couple of days to see if we like it.

Reply:
cc @Maestro31

Reply:
As an idea of Abstraction, maybe we like to have a 'Runner' class instead of a 'Process' one? (maybe runner is not the perfect metaphor...)

Reply:
go for steps -> acts (and step -> acts)